### PR TITLE
[feat]: [CDS-78611]: Passed the actual expression used by user in custom trigger.

### PR DESCRIPTION
**Pull request by senjucanon2**
## Describe your changes
Context: For Custom triggers, default value <+trigger.branch> is not at all used at BE, instead <+trigger.payload.branch> is used. Also If user uses his expression in the trigger, that is also not honoured.

Change: Have passed the actual expression passed by user for branch expression evaluation. But for backward incompatibility, support for <+trigger.branch> is also kept. UI should not fill this value by default, instead <+trigger.payload.branch> is valid.

Impact: Custom Triggers for remote pipelines.

Testing: Have tested with customer expression with the custom payload. Pipeline was executed. Also previous support for <+trigger.branch> works as before.

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52565)

### CODE REVIEW

**Comment by senjucanon2** [comment]
trigger messagecheck

File: pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java [hunk 1]
```diff
@@ -518,8 +519,14 @@ public String resolveBranchExpression(String expression, TriggerWebhookEvent tri
       } else {
         triggerExpressionEvaluator = WebhookTriggerFilterUtils.generatorPMSExpressionEvaluator(
             null, triggerWebhookEvent.getHeaders(), triggerWebhookEvent.getPayload());
+
+        String branchExpression = expression;
+        // For Backward Compatibility Trigger Branch is also handled.
+        if (EmptyPredicate.isEmpty(branchExpression) || TRIGGER_BRANCH.equals(branchExpression)) {
```
Comment (Y0709) [inline pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java:525]: Can we add a UT for this case

**Review by Y0709 — CHANGES_REQUESTED** [review]

**Review by Y0709 — APPROVED** [review]

**Comment by senjucanon2** [comment]
trigger uts

**Review by vcalasansh — COMMENTED** [review]
Thanks man! Change looks good to me and this was long pending item to be fixed.
But what do you say we put this behind a Feature Flag? I'm afraid we could cause unexpected change for some customer who might have saved his trigger with a custom expression in `pipelineBranchName` and figured the only way to make it work was to add a `branch` key in the payload. If this is the case for someone, it is possible they have to fix the payload they are sending in the trigger. With a Feature Flag it would be easy to rollback until customer can adapt.

Also, please make sure to add this in the release notes.

File: pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java [hunk 2]
```diff
@@ -518,8 +519,14 @@ public String resolveBranchExpression(String expression, TriggerWebhookEvent tri
       } else {
         triggerExpressionEvaluator = WebhookTriggerFilterUtils.generatorPMSExpressionEvaluator(
             null, triggerWebhookEvent.getHeaders(), triggerWebhookEvent.getPayload());
+
```
Comment (sribalij) [inline pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java:522]: Can we use a FF just to make sure

**Comment by senjucanon2** [comment]
trigger commonchecks, uts

**Review by senjucanon2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52565)* status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java` line 525 at r1](https://reviewable.io/reviews/harness/harness-core/52565#gh-1318143788:-Ne24IZ_2tN_Sfq8-fa8:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8f7931dd704b4749f14d9443394d7aba12fa5745/pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java#L525)):*
<details><summary><i>Previously, Y0709 (Yuvraj Singhal) wrote…</i></summary><blockquote>

Can we add a UT for this case
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java` line 522 at r2](https://reviewable.io/reviews/harness/harness-core/52565#gh-1320232677:-Ne24IHJAgZC63FjvcD4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/622152f8aa167cfa3e2fa2ce2f23d89f08cd29fe/pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java#L522)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

Can we use a FF just to make sure
</blockquote></details>

Done.

**Comment by senjucanon2** [comment]
trigger uts

**Comment by senjucanon2** [comment]
trigger ut0, ut3, ut4, ut9

**Review by vcalasansh — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52565)* status: 4 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by Y0709 — DISMISSED** [review]

File: pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java [hunk 3]
```diff
@@ -553,8 +557,15 @@ public String resolveBranchExpression(String expression, TriggerWebhookEvent tri
...
+        String branchExpression = TRIGGER_PAYLOAD_BRANCH;
+        // For Backward Compatibility Trigger Branch is also handled.
+        if (pmsFeatureFlagHelper.isEnabled(triggerWebhookEvent.getAccountId(), RESOLVE_CUSTOM_TRIGGER_EXPRESSION)
```
Comment (sribalij) [inline pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java:563]: Can you add a CDS label for the FF? may be something like CDS_RESOLVE_CUSTOM_TRIGGER_EXPRESSION

Comment (senjucanon2) [reply pipeline-service/service/src/main/java/io/harness/pms/triggers/TriggerExecutionHelper.java]: Done

**Comment by senjucanon2** [comment]
trigger uts

**Comment by senjucanon2** [comment]
trigger uts

**Comment by senjucanon2** [comment]
trigger uts

**Review by Y0709 — APPROVED** [review]

**Review by sribalij — APPROVED** [review]

**Comment by senjucanon2** [comment]
trigger review

**Comment by senjucanon2** [comment]
trigger ut0, ut3, ut4, ut9

**Comment by senjucanon2** [comment]
trigger uts

**Review by Y0709 — APPROVED** [review]

**Review by ankittiwari-harness — APPROVED** [review]

**Pull request merged**